ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp

```diff
@@ -196,17 +197,19 @@ Tensor all_gather(
     if (num_devices == 1){
         topology = all_gather_op::Topology::Linear;
     }
+    auto mesh_device = MeshDevice::fetch_mesh_device(devices);
+    auto ring_devices = mesh_device->get_ring_devices();
```

Comment:
This line is really my only current concern.

I agree that we should have only one allgather entrypoint which can temporarily be qualified with a topology. Then allgather can request a chip order  accordingly.

However, I wouldn't want the op to have to request a ring or line but instead just be given a sequence of chips and it decides if it can make a ring out of it or not. This way we get portability across galaxy variants ( those that are and aren't torused up) and the op can automatically take advantage of the torus without the model writer having to update all their ccl calls.

This is more inline with allgather forcing a line for a 2 chip invocation.

Reply:
> I agree that we should have only one allgather entrypoint which can temporarily be qualified with a topology. Then allgather can request a chip order accordingly. I also agree we need a way to have a consistent way to get an order of chips.

> However, I wouldn't want the op to have to request a ring or line but instead just be given a sequence of chips and it decides if it can make a ring out of it or not.

Seems to conflict. MeshTensor is mapped onto a set of devices, and invocation to `ttnn.all_gather(mesh_tensor)` will land here. What we're doing here is requesting the ring devices from the mesh tensor.